LiS/man/man4/mtdrv.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: mtdrv.4.man,v 0.9.2.3 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: mtdrv.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH XXXX 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B xxx
\- incomplete man page
.\"
.\"
.SH SYNOPSIS
.PP
none
.\"
.\"
.SH DESCRIPTION
.PP
This manpage is under construction.
.\"
.\"
.SH FILES
.PP
.\"
.\"
.SH "SEE ALSO"
.PP
.\"
.\"
.SH BUGS
.PP
Not done yet.
.\"
.\"
.[
$LIST$
.]
.\"
.\"
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/linux/mtdrv.c
/*****************************************************************************

 @(#) $RCSfile: mtdrv.c,v $ $Name:  $($Revision: 1.1.1.3.4.14 $) $Date: 2008-04-29 08:33:12 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:12 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: mtdrv.c,v $
 Revision 1.1.1.3.4.14  2008-04-29 08:33:12  brian
 - update headers for Affero release

 Revision 1.1.1.3.4.13  2007/08/15 04:58:03  brian
 - GPLv3 updates

 *****************************************************************************/

#ident "@(#) $RCSfile: mtdrv.c,v $ $Name:  $($Revision: 1.1.1.3.4.14 $) $Date: 2008-04-29 08:33:12 $"

static char const ident[] = "$RCSfile: mtdrv.c,v $ $Name:  $($Revision: 1.1.1.3.4.14 $) $Date: 2008-04-29 08:33:12 $";

/************************************************************************
*                     Multi-Threaded Test Driver                        *
*************************************************************************
*									*
* Used to test the multi-threaded aspects of driver opens.  This is	*
* a separately loadable driver which is only loaded when strtst is run.	*
*									*
************************************************************************/

/*  -------------------------------------------------------------------  */

#include <sys/LiS/module.h>	/* must be VERY first include */

#include <sys/stream.h>
#include <sys/stropts.h>
#include <linux/stat.h>		/* for S_IFCHR */
#include <sys/osif.h>

#include <sys/LiS/mtdrv.h>

/************************************************************************
*                        Module Definition Structs                      *
************************************************************************/

static struct module_info mtdrv_minfo = {
	0,				/* id */
	"mtdrv",			/* name */
	0,				/* min packet size accepted */
	INFPSZ,				/* max packet size accepted */
	10240L,				/* high water mark */
	512L				/* low water mark */
};

/* These are the entry points to the driver: open, close, write side put and
 * service procedures and read side service procedure.
 */
static int _RP mtdrv_open(queue_t *, dev_t *, int, int, cred_t *);
static int _RP mtdrv_close(queue_t *, int, cred_t *);
static int _RP mtdrv_wput(queue_t *, mblk_t *);
static int _RP mtdrv_wsrv(queue_t *);
static int _RP mtdrv_rsrv(queue_t *);

/* qinit structures (rd and wr side) 
 */
static struct qinit mtdrv_rinit = {
	NULL,				/* put */
	mtdrv_rsrv,			/* service */
	mtdrv_open,			/* open */
	mtdrv_close,			/* close */
	NULL,				/* admin */
	&mtdrv_minfo,			/* info */
	NULL				/* stat */
};

static struct qinit mtdrv_winit = {
	mtdrv_wput,			/* put */
	mtdrv_wsrv,			/* service */
	NULL,				/* open */
	NULL,				/* close */
	NULL,				/* admin */
	&mtdrv_minfo,			/* info */
	NULL				/* stat */
};

/* streamtab for the driver.
 */
struct streamtab mtdrv_info = {
	&mtdrv_rinit,			/* read queue */
	&mtdrv_winit,			/* write queue */
	NULL,				/* mux read queue */
	NULL				/* mux write queue */
};

/************************************************************************
*                        Global Storage                                 *
************************************************************************/

static int next_clone_dev;
static int open_sleep_time;
static lis_semaphore_t *open_sleep_sem;
static int mtdrv_major;			/* from registration */

static lis_semaphore_t *open_sem;

#define N_DEVS	20
static char devs[N_DEVS];

/************************************************************************
*                           Prototypes                                  *
************************************************************************/

static void _RP mtdrv_timeout(caddr_t arg);

/************************************************************************
*                          Functions                                    *
************************************************************************/

static void
make_node(char *name, int maj, int mnr)
{
	int rtn;

	rtn = lis_unlink(name);
	if (rtn < 0 && rtn != -ENOENT) {
		printk("unlink %s: error %d\n", name, rtn);
		return;
	}

	if (maj == 0) {		/* don't make the node */
		printk("unlink %s: OK\n", name);
		return;
	}

	rtn = lis_mknod(name, 0666 | S_IFCHR, UMKDEV(maj, mnr));
	if (rtn == 0)
		printk("mknod %s: OK\n", name);
	else
		printk("mknod %s: error %d\n", name, rtn);
}

void
mtdrv_init(int maj)
{
	make_node("/dev/mtdrv_clone", CLONE__CMAJOR_0, maj);
	make_node("/dev/mtdrv.1", maj, 1);
	make_node("/dev/mtdrv.2", maj, 2);
	make_node("/dev/mtdrv.3", maj, 3);
	make_node("/dev/mtdrv.4", maj, 4);
	open_sleep_sem = lis_sem_alloc(0);
	open_sem = lis_sem_alloc(1);

}				/* mtdrv_init */

void
mtdrv_term(void)
{
	open_sleep_sem = lis_sem_destroy(open_sleep_sem);
	open_sem = lis_sem_destroy(open_sem);
	make_node("/dev/mtdrv_clone", 0, 0);
	make_node("/dev/mtdrv.1", 0, 0);
	make_node("/dev/mtdrv.2", 0, 0);
	make_node("/dev/mtdrv.3", 0, 0);
	make_node("/dev/mtdrv.4", 0, 0);

}				/* mtdrv_term */

/*  -------------------------------------------------------------------  */

static int _RP
mtdrv_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *credp)
{
	int dev;

	printk("mtdrv_open: flag=0x%x sflag=0x%x\n", flag, sflag);
	lis_down(open_sem);
	if (sflag == CLONEOPEN) {
		if (next_clone_dev) {
			dev = next_clone_dev;
			next_clone_dev = 0;
			if (dev >= sizeof(devs)) {
				lis_up(open_sem);
				return (-EINVAL);
			}
		} else {
			for (dev = 1; dev < sizeof(devs); dev++)
				if (devs[dev] == 0)
					break;

			if (dev >= sizeof(devs)) {
				lis_up(open_sem);
				return (-EBUSY);
			}

		}
	} else
		dev = getminor(*devp);

	MODGET();
	devs[dev]++;		/* must be protected by semaphore */

	/* 
	 * Note: we may be forced by an ioctl to "select" an already open
	 * dev as a clone device to test an error condition.  In that case
	 * we do not want to set the q_ptr to the devs array if that dev
	 * is already in use.  We will get a close call for the erroneous
	 * open and we want the pointer to be NULL when close is called.
	 */
	if (devs[dev] == 1)	/* first open */
		q->q_ptr = (void *) &devs[dev];
	lis_up(open_sem);
	*devp = makedevice(getmajor(*devp), dev);

	printk("mtdrv_open: dev=%d\n", dev);
	if (open_sleep_time > 0) {
		timeout(mtdrv_timeout, (caddr_t) 0, open_sleep_time);
		open_sleep_time = 0;
		lis_down(open_sleep_sem);
	}

	return 0;		/* success */
}

/*  -------------------------------------------------------------------  */

static void _RP
mtdrv_timeout(caddr_t arg)
{
	lis_up(open_sleep_sem);
}				/* mtdrv_timeout */

/*  -------------------------------------------------------------------  */

static void
mtdrv_ioctl(queue_t *q, mblk_t *mp)
{
	struct iocblk *iocb;
	int error;
	int rtn_count = 0;

	iocb = (struct iocblk *) mp->b_rptr;

	/* 
	 * Syntax switch
	 */
	switch (iocb->ioc_cmd) {
	case MTDRV_SET_CLONE_DEV:
	case MTDRV_SET_OPEN_SLEEP:
		if (mp->b_cont == NULL) {
			printk("mtdrv_ioctl: integer argument expected\n");
			error = EINVAL;
			goto iocnak;
		}
		if (iocb->ioc_count != sizeof(int)) {
			printk("mtdrv_ioctl: expected ioctl len %ld, got %ld\n", (long) sizeof(int),
			       (long) iocb->ioc_count);

			error = EINVAL;
			goto iocnak;
		}
		break;

	default:
		printk("mtdrv_ioctl: Invalid ioctl: %x\n", iocb->ioc_cmd);
		error = EINVAL;
		goto iocnak;

	      iocnak:
		{
			mp->b_datap->db_type = M_IOCNAK;
			iocb->ioc_error = error;
			qreply(q, mp);
		}
		return;
	}			/* end of switch on ioc_cmd */

	/* 
	 * Semantics switch
	 */
	switch (iocb->ioc_cmd) {
	case MTDRV_SET_CLONE_DEV:
		next_clone_dev = *(int *) mp->b_cont->b_rptr;
		break;
	case MTDRV_SET_OPEN_SLEEP:
		open_sleep_time = *(int *) mp->b_cont->b_rptr;
		break;

	}			/* end of switch on ioc_cmd */

	mp->b_datap->db_type = M_IOCACK;
	iocb->ioc_count = rtn_count;
	iocb->ioc_error = 0;
	qreply(q, mp);
}

/*  -------------------------------------------------------------------  */

static int _RP
mtdrv_wput(queue_t *q, mblk_t *mp)
{

	switch (mp->b_datap->db_type) {
	case M_IOCTL:
		mtdrv_ioctl(q, mp);
		return (0);

	case M_FLUSH:
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
		}
		if (mp->b_rptr[0] & FLUSHR) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(RD(q), mp->b_rptr[1], FLUSHDATA);
			else
				flushq(RD(q), FLUSHDATA);
			mp->b_rptr[0] &= ~FLUSHW;
			qreply(q, mp);
		} else
			freemsg(mp);
		return (0);

	default:
		break;
	}

	freemsg(mp);
	return (0);
}

/*  -------------------------------------------------------------------  */

static int _RP
mtdrv_wsrv(queue_t *q)
{
	return (0);
}

/*  -------------------------------------------------------------------  */

static int _RP
mtdrv_rsrv(queue_t *q)
{
	return (0);
}

/*  -------------------------------------------------------------------  */

static int _RP
mtdrv_close(queue_t *q, int dummy, cred_t *credp)
{
	char *devp = (char *) q->q_ptr;

	while (devp != NULL && *devp) {
		MODPUT();
		(*devp)--;
	}
	return (0);
}

#if !defined __NO_VERSION__

#ifdef KERNEL_2_5
int
mtdrv_init_module(void)
#else
int
init_module(void)
#endif
{
	int ret = lis_register_strdev(0, &mtdrv_info, 10, "mtdrv");

	if (ret < 0) {
		printk("mtdrv.init_module: Unable to register driver.\n");
		return ret;
	}
	mtdrv_major = ret;
	mtdrv_init(mtdrv_major);
	return 0;
}

#ifdef KERNEL_2_5
void
mtdrv_cleanup_module(void)
#else
void
cleanup_module(void)
#endif
{
	mtdrv_term();
	if (lis_unregister_strdev(mtdrv_major) < 0)
		printk("mtdrv.cleanup_module: Unable to unregister driver.\n");
	else
		printk("mtdrv.cleanup_module: Unregistered, ready to be unloaded.\n");
	return;
}

#ifdef KERNEL_2_5
module_init(mtdrv_init_module);
module_exit(mtdrv_cleanup_module);
#endif

#if defined(MODULE_LICENSE)
MODULE_LICENSE("GPL");
#endif
#if defined(MODULE_AUTHOR)
MODULE_AUTHOR("David Grothe");
#endif
#if defined(MODULE_DESCRIPTION)
MODULE_DESCRIPTION("multi-threaded open test helper driver");
#endif
#if defined(MODULE_ALIAS)
MODULE_ALIAS("streams-" __stringify(LIS_OBJNAME));
#endif

#endif

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0));
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-*");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-1");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-2");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-3");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-4");
MODULE_ALIAS("/dev/mtdrv_clone");
MODULE_ALIAS("/dev/mtdrv.1");
MODULE_ALIAS("/dev/mtdrv.2");
MODULE_ALIAS("/dev/mtdrv.3");
MODULE_ALIAS("/dev/mtdrv.4");
#endif
